README.md
# Kuwait Prayer Times
Kuwait Prayer Times based on [AlAdhan](https://aladhan.com) API services

[Amiri font](https://www.amirifont.org) used for Arabic text

## Getting Started
Include our JavaScript before the closing body

`<script src="https://q8p.io/js"></script>`

Include our CSS in your project

`<link rel="stylesheet" href="https://q8p.io/css">`

Or by using Node.js

`npm i q8-prayertimes`

`<script src="./node_modules/q8-prayertimes/dist/all.min.js"></script>`

`<link rel="stylesheet" href="./node_modules/q8-prayertimes/dist/all.min.css">`

Create a div element with an id of q8prayers in your project

`<div id="q8prayers"></div>`

> You can add more than one element in a single page

## Options
- Languages
  - `data-lang="en"` Display prayer times in English *(default)*
  - `data-lang="ar"` Display prayer times in Arabic
- Date Types:
  - `data-date="gregorian"` Display date in Gregorian format *(default)*
  - `data-date="hijri"` Display date in Hijri format
- Themes:
  - `data-theme="light"` Display prayer times in ligh mode *(default)*
  - `data-theme="dark"` Display prayer times in dark mode
  - `data-theme="black"` Display prayer times in all black mode
  - `data-theme="white"` Display prayer times in all white mode
- Table Sizes:
  - `data-size="sm"` Display table in small size
  - `data-size="md"` Display table in medium size *(default)*
  - `data-size="lg"` Display table in large size
  - `data-size="xl"` Display table in xlarge size

## Starter Template
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <!-- Required meta tags -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <!-- Q8 Prayer Times CSS -->
    <link rel="stylesheet" href="https://q8p.io/css">

    <title>Q8 Prayer Times Starter Template</title>
  </head>
  <body>
    <!-- Q8 Prayer Times Element -->
    <div id="q8prayers" data-lang="ar"></div>

    <!-- Q8 Prayer Times JavaScript -->
    <script src="https://q8p.io/js"></script>
  </body>
</html>
```

## Hint
Add the following styling to center prayer times table in a `<div>` element

```css
#q8prayers table{
  margin-right: auto;
  margin-left: auto;
}
```


CHANGELOG.md
# Changelog

## 2.0.0 (2022-1-7)
### Changed
* Table border radius set to .75 rem

## 2.0.0-rc.1 (2022-1-2)
### Added
* Date types option, choose wheather Gregorian or Hijri date format

### Changed
* Table header is the credit now
* Table style

## 2.0.0-rc.0 (2022-1-1)
### Changed
* Fetch from [q8prayers API](https://github.com/m2kio/q8-prayers-api) instead of direct [AlAdhan API](https://aladhan.com)
* Table width increased for `sm` (150px) and `md` (175px) table sizes

### Deleted
* Width option
* reformatDate function
* toArabicNumbers function

## 1.3.1 (2021-12-5)
### Fixed
* `<tr>` padding
* `<td>` & `<th>` font size

## 1.3.0 (2021-11-18)
### Added
* Table sizing feature

## 1.2.1 (2021-11-17)
### Changed
* Font sizes from rem to em

## 1.2.0 (2021-10-20)
### Added
* All black theme
* All white theme

### Changed
* Using CSS variables instead of SASS variables

## 1.1.0 (2021-10-19)
### Deleted
* CSS import functionality

## 1.0.12 (2021-9-2)
### Fixed
* gh link

## 1.0.11 (2021-5-7)
### Fixed
* AlAdhan API endpoint

## 1.0.10 (2021-2-6)
### Fixed
* ampm format fixed

## 1.0.9 (2021-1-21)
### Fixed
* Multiple requests when importing more than one `<div>` tag

## 1.0.8 (2021-1-20)
### Changed
* Credits changed to https://q8prayers.com

## 1.0.7 (2021-1-19)
### Added
* CSS load script
* Overpass font

### Deleted
* Main JS source build

## 1.0.6 (2021-1-18)
### Changed
* Homepage links to instruction page in package.json
* Demo link changed to instructions
* Table footer links to instruction page

### Deleted
* Demo link from README

## 1.0.5 (2021-1-17)
### Added
* Multiple elements in a single page support

## 1.0.4 (2021-1-13)
### Added
* Repository to package.json
* Email to package.json
* License to header banner
* Darkmode support

### Changed
* SASS use variables

### Fixed
* Table font color

## 1.0.3 (2021-1-3)
### Added
* Amiri font credits in README
* Kuwait Prayer Times credits in prayers table

### Changed
* jsDelivr omit version from demo page and README

## 1.0.2 (2020-12-29)
### Added
* Demo link in README
* Title at the top of the table

## 1.0.1 (2020-12-28)
### Added
* node package dev and build scripts added

### Changed
* Prayer times table text alignment corrected
* Error handled

### Removed
* Banner header from dev sass and js

## 1.0.0 (2020-12-27)
### Added
* Everything


js/app.js
const API_URL = 'https://api.q8prayers.com'
const prayerElements = document.querySelectorAll('#q8prayers')

if (prayerElements) {

  fetch(API_URL).then(res => res.json())
  .then(data => {
    if (data.status === 200) { 
      let lang, width, fontSize, table, tableTitle, tableDate

      prayerElements.forEach(prayerElement => {
        prayerElement.getAttribute('data-lang') === 'ar' ? lang = 'ar' : lang = 'en'
        const tableSize = prayerElement.getAttribute('data-size')
        tableDate = prayerElement.getAttribute('data-date')
        tableDate != null ? tableDate = tableDate.toLowerCase() : tableDate = null

        // detect table size
        if (tableSize !== null) {
          if (tableSize === 'sm'){
            fontSize = 10
            width = 150
          }
          if (tableSize === 'md'){
            fontSize = 15
            width = 175
          }
          if (tableSize === 'lg'){
            fontSize = 20
            width = 200
          }
          if (tableSize === 'xl'){
            fontSize = 25
            width = 250
          }
        } else {
          fontSize = 15
          width = 175
        }

        // prayers variables
        const FajrName = data.timings.Fajr[lang].name
        const FajrTime = data.timings.Fajr[lang].time
        const DhuhrName = data.timings.Dhuhr[lang].name
        const DhuhrTime = data.timings.Dhuhr[lang].time
        const AsrName = data.timings.Asr[lang].name
        const AsrTime = data.timings.Asr[lang].time
        const MaghribName = data.timings.Maghrib[lang].name
        const MaghribTime = data.timings.Maghrib[lang].time
        const IshaName = data.timings.Isha[lang].name
        const IshaTime = data.timings.Isha[lang].time
        const Gdate = data.date.gregorian[lang].date
        const Hdate = data.date.hijri[lang].date

        // configure date type
        tableDate === 'hijri' ? tableDate = Hdate : tableDate = Gdate

        // main table configuration
        table = document.createElement('table')
        table.style.width = `${width}px`
        table.style.fontSize = `${fontSize}px`
        if(lang === 'ar') table.setAttribute('dir', 'rtl')
        lang === 'ar' ? tableTitle = 'أوقات الصلاة في الكويت' : tableTitle = 'Kuwait Prayer Times'
  
        // table data
        table.innerHTML = `<tbody>
          <tr><th colspan="2"><small><a href="https://q8p.io" target="_blank">${tableTitle}</a></small></th></tr>
          <tr><th colspan="2">${tableDate}</th></tr>
          <tr><th>${FajrName}</th><td>${FajrTime}</td></tr>
          <tr><th>${DhuhrName}</th><td>${DhuhrTime}</td></tr>
          <tr><th>${AsrName}</th><td>${AsrTime}</td></tr>
          <tr><th>${MaghribName}</th><td>${MaghribTime}</td></tr>
          <tr><th>${IshaName}</th><td>${IshaTime}</td></tr>
          </tbody>`
    
        // append table to DOM
        prayerElement.append(table)
      })

    } else {
      prayerElements.forEach(prayerElement => {
        prayerElement.innerHTML = `<p class="q8prayers-error">Q8Prayers Internal Error</p>`
      })
    }
  })

  .catch(err => {
    console.log(err)
    prayerElements.forEach(prayerElement => {
      prayerElement.innerHTML = `<p class="q8prayers-error">Q8Prayers Internal Error</p>`
    })
  })
  
}


gulpfile.js
var gulp = require('gulp')
var sass = require('gulp-sass')
var uglifycss = require('gulp-uglifycss')
var rename = require("gulp-rename")
var concat = require('gulp-concat')
var ignore = require('gulp-ignore')
var gutil = require('gulp-util')
var minifyJS = require('gulp-minify')
var header = require("gulp-header")
var pkg = require("./package.json")

var banner = ["/**",
    " * <%= pkg.name %> v<%= pkg.version %>",
	" * @link <%= pkg.homepage %>",
	" * @repo <%= pkg.repository.url %>",
	" *",
	" * <%= pkg.description %>",
	" * Author <%= pkg.author.name %> <<%= pkg.author.email %>>",
	" * License <%= pkg.license %>",
	" */",
  ""].join("\n")

// development tasks
gulp.task('scripts-dev', async function(){
  gulp.src('./js/**/*.js')
  .pipe(concat('dev.js'))
  .pipe(ignore.exclude(['**/*.map']))
  .on('error', function(err){ gutil.log(gutil.colors.red('[Error]'), err.toString()) })
  .pipe(gulp.dest('./dev'))
})

gulp.task('sass-dev', async function(){
  gulp.src('./scss/**/*.scss')
    .pipe(sass().on('error', sass.logError))
    .pipe(rename("dev.css"))
    .pipe(gulp.dest('./dev'))
})

// production tasks
gulp.task('scripts', async function(){
  gulp.src('./js/**/*.js')
  .pipe(concat('all.js'))
  .pipe(ignore.exclude(['**/*.map']))
  .pipe(minifyJS({
    ext:{
      min:'.min.js'
    },
    noSource: true
  }))
  .on('error', function(err){ gutil.log(gutil.colors.red('[Error]'), err.toString()) })
  .pipe(header(banner, {pkg: pkg}))
  .pipe(gulp.dest('./dist'))
})

gulp.task('sass', async function(){
  gulp.src('./scss/**/*.scss')
    .pipe(sass().on('error', sass.logError))
    .pipe(uglifycss())
    .pipe(rename("all.min.css"))
    .pipe(header(banner, {pkg: pkg}))
    .pipe(gulp.dest('./dist'))
})

// tasks
gulp.task('run-dev', gulp.series('sass-dev', 'scripts-dev'))

gulp.task('watch', function(){
  gulp.watch('./scss/**/*.scss', gulp.series('sass-dev'))
  gulp.watch('./js/**/*.js', gulp.series('scripts-dev'))
})

gulp.task('build', gulp.series('sass', 'scripts'))
gulp.task('default', gulp.series('run-dev', 'watch'))


js/fetchPrayers.js
async function fetchPrayers(today){
  try {
    const res = await (await fetch(`https://api.aladhan.com/v1/timingsByAddress/${today}?address=kuwait,al-asimah&method=9`)).json()
    if (res.code === 200) {
      const Fajr = res.data.timings.Fajr
      const Dhuhr = res.data.timings.Dhuhr
      const Asr = res.data.timings.Asr
      const Maghrib = res.data.timings.Maghrib
      const Isha = res.data.timings.Isha
  
      const prayers = {
        Fajr: {
          en: {
            name: 'Fajr',
            time: `${reformatDate('en', Fajr)}`
          },
          ar: {
            name: 'الفجر',
            time: `${reformatDate('ar', Fajr).toArabicNumbers()}`
          }
        },
        Dhuhr: {
          en: {
            name: 'Dhuhr',
            time: `${reformatDate('en', Dhuhr)}`
          },
          ar: {
            name: 'الظُهر',
            time: `${reformatDate('ar', Dhuhr).toArabicNumbers()}`
          }
        },
        Asr: {
          en: {
            name: 'Asr',
            time: `${reformatDate('en', Asr)}`
          },
          ar: {
            name: 'العصر',
            time: `${reformatDate('ar', Asr).toArabicNumbers()}`
          }
        },
        Maghrib: {
          en: {
            name: 'Maghrib',
            time: `${reformatDate('en', Maghrib)}`
          },
          ar: {
            name: 'المغرب',
            time: `${reformatDate('ar', Maghrib).toArabicNumbers()}`
          }
        },
        Isha: {
          en: {
            name: 'Isha',
            time: `${reformatDate('en', Isha)}`
          },
          ar: {
            name: 'العِشاء',
            time: `${reformatDate('ar', Isha).toArabicNumbers()}`
          }
        },
        Date: {
          en: today,
          ar: today.toArabicNumbers()
        },
        Title: {
          en: 'Kuwait Prayer Times',
          ar: 'أوقات الصلاة في الكويت'
        }
      }
      
      return prayers

    }else{
      const message = {
        en: 'Please try again later',
        ar: 'الرجاء المحاولة لاحقًا'
      }
      return { error: true, message }
    }

  } catch (error) {
    console.error(error)
    const message = {
      en: 'Please try again later',
      ar: 'الرجاء المحاولة لاحقًا'
    }
    return { error: true, message }
  }
}


js/dev.js
String.prototype.toArabicNumbers= function(){
  var id= ['٠', '١', '٢', '٣', '٤' ,'٥', '٦', '٧', '٨', '٩']
  return this.replace(/[0-9]/g, function(w){
    return id[+w]
  })
}

function reformatDate(lang, date) {
  const h = date.toString().split(':')[0]
  const m = date.toString().split(':')[1]

  var am, pm, ampm

  lang === 'ar' ? am = 'ص' : am = 'AM'
  lang === 'ar' ? pm = 'م' : pm = 'PM'
  ampm = parseInt(h) >= 12 ? pm : am

  return parseInt(h) > 12 ? `${parseInt(h) - 12}:${m} ${ampm}` : `${parseInt(h)}:${m} ${ampm}`
}

async function fetchPrayers(today){
  try {
    const res = await (await fetch(`https://api.aladhan.com/v1/timingsByAddress/${today}?address=kuwait,al-asimah&method=9`)).json()
    if (res.code === 200) {
      const Fajr = res.data.timings.Fajr
      const Dhuhr = res.data.timings.Dhuhr
      const Asr = res.data.timings.Asr
      const Maghrib = res.data.timings.Maghrib
      const Isha = res.data.timings.Isha
  
      const prayers = {
        Fajr: {
          en: {
            name: 'Fajr',
            time: `${reformatDate('en', Fajr)}`
          },
          ar: {
            name: 'الفجر',
            time: `${reformatDate('ar', Fajr).toArabicNumbers()}`
          }
        },
        Dhuhr: {
          en: {
            name: 'Dhuhr',
            time: `${reformatDate('en', Dhuhr)}`
          },
          ar: {
            name: 'الظُهر',
            time: `${reformatDate('ar', Dhuhr).toArabicNumbers()}`
          }
        },
        Asr: {
          en: {
            name: 'Asr',
            time: `${reformatDate('en', Asr)}`
          },
          ar: {
            name: 'العصر',
            time: `${reformatDate('ar', Asr).toArabicNumbers()}`
          }
        },
        Maghrib: {
          en: {
            name: 'Maghrib',
            time: `${reformatDate('en', Maghrib)}`
          },
          ar: {
            name: 'المغرب',
            time: `${reformatDate('ar', Maghrib).toArabicNumbers()}`
          }
        },
        Isha: {
          en: {
            name: 'Isha',
            time: `${reformatDate('en', Isha)}`
          },
          ar: {
            name: 'العِشاء',
            time: `${reformatDate('ar', Isha).toArabicNumbers()}`
          }
        },
        Date: {
          en: today,
          ar: today.toArabicNumbers()
        },
        Title: {
          en: 'Kuwait Prayer Times',
          ar: 'أوقات الصلاة في الكويت'
        }
      }
      
      return prayers

    }else{
      const message = {
        en: 'Please try again later',
        ar: 'الرجاء المحاولة لاحقًا'
      }
      return { error: true, message }
    }

  } catch (error) {
    console.error(error)
    const message = {
      en: 'Please try again later',
      ar: 'الرجاء المحاولة لاحقًا'
    }
    return { error: true, message }
  }
}

let today = new Date()
const dd = String(today.getDate()).padStart(2, '0')
const mm = String(today.getMonth() + 1).padStart(2, '0')
const yyyy = today.getFullYear()
today = dd + '-' + mm + '-' + yyyy

const prayerElements = document.querySelectorAll('#q8prayers')

fetchPrayers(today).then(prayers => {

  prayerElements.forEach(prayerElement => {
    var lang, width, dark
  
    prayerElement.getAttribute('data-lang') === 'ar' ? lang = 'ar' : lang = 'en'
    prayerElement.getAttribute('data-width') === null ? width = 150 : width = parseInt(prayerElement.getAttribute('data-width'))
    prayerElement.getAttribute('data-theme') === 'dark' ? dark = true : dark = false

    if (prayers.error) {
      prayerElement.innerHTML = `<p class="q8prayers-error">${lang === 'ar' ? prayers.message.ar : prayers.message.en}</p>`
      
    }else{
      if(dark){
        prayerElement.classList.add('dark')
      }
      
      let table = document.createElement('table')
      if(lang === 'ar') table.setAttribute('dir', 'rtl')
    
      let title_tr = document.createElement('tr')
      let title_th = document.createElement('th')
      let title_small = document.createElement('small')
  
      let credits_tr = document.createElement('tr')
      let credits_th = document.createElement('th')
      let credits_small = document.createElement('small')
      let credits_a = document.createElement('a')
  
      let date_tr = document.createElement('tr')
      let date_th = document.createElement('th')
    
      let fajr_tr = document.createElement('tr')
      let fajr_th = document.createElement('th')
      let fajr_td = document.createElement('td')
    
      let dhuher_tr = document.createElement('tr')
      let dhuher_th = document.createElement('th')
      let dhuher_td = document.createElement('td')
    
      let asr_tr = document.createElement('tr')
      let asr_th = document.createElement('th')
      let asr_td = document.createElement('td')
    
      let maghrib_tr = document.createElement('tr')
      let maghrib_th = document.createElement('th')
      let maghrib_td = document.createElement('td')
    
      let isha_tr = document.createElement('tr')
      let isha_th = document.createElement('th')
      let isha_td = document.createElement('td')
    
      title_th.setAttribute('colspan', '2')
      title_small.innerHTML = lang === 'ar' ? prayers.Title.ar : prayers.Title.en
      title_th.appendChild(title_small)
      title_tr.appendChild(title_th)
    
      credits_th.setAttribute('colspan', '2')
      credits_a.setAttribute('href', 'https://q8prayers.com')
      credits_a.setAttribute('target', '_blank')
      credits_a.innerHTML = 'Q8 Prayer Times'
      credits_small.appendChild(credits_a)
      credits_th.appendChild(credits_small)
      credits_tr.appendChild(credits_th)
  
      date_th.setAttribute('colspan', '2')
      date_th.innerHTML = lang === 'ar' ? prayers.Date.ar : prayers.Date.en
      date_tr.appendChild(date_th)
     
      fajr_th.innerHTML = lang === 'ar' ? prayers.Fajr.ar.name : prayers.Fajr.en.name
      fajr_td.innerHTML = lang === 'ar' ? prayers.Fajr.ar.time : prayers.Fajr.en.time
      fajr_tr.appendChild(fajr_th)
      fajr_tr.appendChild(fajr_td)
     
      dhuher_th.innerHTML = lang === 'ar' ? prayers.Dhuhr.ar.name : prayers.Dhuhr.en.name
      dhuher_td.innerHTML = lang === 'ar' ? prayers.Dhuhr.ar.time : prayers.Dhuhr.en.time
      dhuher_tr.appendChild(dhuher_th)
      dhuher_tr.appendChild(dhuher_td)
     
      asr_th.innerHTML = lang === 'ar' ? prayers.Asr.ar.name : prayers.Asr.en.name
      asr_td.innerHTML = lang === 'ar' ? prayers.Asr.ar.time : prayers.Asr.en.time
      asr_tr.appendChild(asr_th)
      asr_tr.appendChild(asr_td)
     
      maghrib_th.innerHTML = lang === 'ar' ? prayers.Maghrib.ar.name : prayers.Maghrib.en.name
      maghrib_td.innerHTML = lang === 'ar' ? prayers.Maghrib.ar.time : prayers.Maghrib.en.time
      maghrib_tr.appendChild(maghrib_th)
      maghrib_tr.appendChild(maghrib_td)
     
      isha_th.innerHTML = lang === 'ar' ? prayers.Isha.ar.name : prayers.Isha.en.name
      isha_td.innerHTML = lang === 'ar' ? prayers.Isha.ar.time : prayers.Isha.en.time
      isha_tr.appendChild(isha_th)
      isha_tr.appendChild(isha_td)
    
      table.appendChild(title_tr)
      table.appendChild(date_tr)
      table.appendChild(fajr_tr)
      table.appendChild(dhuher_tr)
      table.appendChild(asr_tr)
      table.appendChild(maghrib_tr)
      table.appendChild(isha_tr)
      table.appendChild(credits_tr)
    
      table.style.width = width + 'px'
    
      prayerElement.appendChild(table)
    }
  })

})
